File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) <= 1:
            return len(s)
        left = 0
        right = 1
        seen = set(s[0])
        longest = 1
        while right < len(s):
            if s[right] in seen:
                while s[right] in seen:
                    seen.remove(s[left])
                    left += 1
            else:
                seen.add(s[right])
                longest = max(longest, right - left + 1)
                right += 1

        return longest

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return goodNodesHelper(root, root.val);
    }

    private int goodNodesHelper(TreeNode root, int max) {
        if (root == null) {
            return 0;
        }

        int newMax = Math.max(root.val, max);
        int curr = 0;

        if (root.val >= max) {
            curr++;
        }

        return curr + goodNodesHelper(root.left, newMax) + goodNodesHelper(root.right, newMax);
    }
}
